File: puzzle_1.py
from enum import Enum

class CompareResult(Enum):
    IN_ORDER = 1
    OUT_OF_ORDER = 2
    INCONCLUSIVE = 3

lines = []

def construct_packet_from_string(s):
    current_list = []
    parent_stack = []
    i = 1
    while i < len(s):
        if i == len(s) - 1:
            return current_list
        elif s[i] == '[':
            new_child_list = []
            parent_stack.append(current_list)
            current_list.append(new_child_list)
            current_list = new_child_list
            i += 1
        elif s[i] == ']':
            current_list = parent_stack.pop()
            i += 1
        elif s[i] == ',':
            i += 1
        else:
            current_number_str = ''
            while s[i].isdigit():
                current_number_str += s[i]
                i += 1
            current_list.append(int(current_number_str))

def convert_to_string(x):
    if type(x) == int:
        return str(x)
    return_string = '['
    for i in range(len(x)):
        return_string += convert_to_string(x[i])
        if i < len(x) - 1:
            return_string += ','
    return_string += ']'
    return return_string

def compare(left, right):
    # print('Compare', convert_to_string(left), 'with', convert_to_string(right))
    if type(left) == int and type(right) == int:
        if left < right:
            # print('Left side is smaller, so inputs are in the right order')
            return CompareResult.IN_ORDER
        elif right < left:
            # print('Right side is smaller, so inputs are not in the right order')
            return CompareResult.OUT_OF_ORDER
        else:
            return CompareResult.INCONCLUSIVE

    elif type(left) == list and type(right) == list:
        for i in range(min(len(left), len(right))):
            result = compare(left[i], right[i])
            if result != CompareResult.INCONCLUSIVE:
                return result
        if len(left) < len(right):
            # print('Left side ran out of items, so inputs are in the right order')
            return CompareResult.IN_ORDER
        elif len(right) < len(left):
            # print('Right side ran out of items, so inputs are not in the right order')
            return CompareResult.OUT_OF_ORDER
        else:
            return CompareResult.INCONCLUSIVE
    
    else:
        if type(left) == int and type(right) == list:
            return compare([left], right)
        else:
            return compare(left, [right])

with open('./input.txt', 'r') as f:
    for line in f:
        lines.append(line.strip())

result = 0

i = 0

pair_index = 1

while i < len(lines):
    first_line = lines[i]
    i += 1
    second_line = lines[i]
    i += 2

    first_packet = construct_packet_from_string(first_line)
    second_packet = construct_packet_from_string(second_line)

    if compare(first_packet, second_packet) == CompareResult.IN_ORDER:
        result += pair_index
    
    pair_index += 1

print(result)

File: Puzzle1.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.Stack;

public class Puzzle1 {
    public static boolean debug = true;

    public static ArrayList<Object> constructListFromString(String s) {
        int i = 1;
        ArrayList<Object> currentList = new ArrayList<>();
        Stack<ArrayList<Object>> parentStack = new Stack<>();

        while (i < s.length() - 1) {
            if (s.charAt(i) == '[') {
                parentStack.push(currentList);
                ArrayList<Object> childList = new ArrayList<>();
                currentList.add(childList);
                currentList = childList;
                i++;
            } else if (s.charAt(i) == ',') {
                i++;
            } else if (s.charAt(i) == ']') {
                currentList = parentStack.pop();
                i++;
            } else {
                String currentNumberString = "";
                while (Character.isDigit(s.charAt(i))) {
                    currentNumberString += s.charAt(i);
                    i++;
                }
                currentList.add(Integer.parseInt(currentNumberString));
            }
        }

        return currentList;
    }

    public static int compare(ArrayList<Object> left, ArrayList<Object> right) {
        for (int i = 0; i < Math.min(left.size(), right.size()); i++) {
            if (left.get(i) instanceof Integer && right.get(i) instanceof Integer) {
                int leftIntValue = ((Integer)left.get(i)).intValue();
                int rightIntValue = ((Integer)right.get(i)).intValue();
                if (leftIntValue < rightIntValue) {
                    return -1;
                } else if (rightIntValue < leftIntValue) {
                    return 1;
                }
            } else if (left.get(i) instanceof ArrayList && right.get(i) instanceof ArrayList) {
                int compareResult = compare((ArrayList<Object>)left.get(i), (ArrayList<Object>)right.get(i));
                if (compareResult != 0) return compareResult;
            } else if (left.get(i) instanceof ArrayList) {
                ArrayList<Object> r = new ArrayList<>();
                r.add(right.get(i));
                int compareResult = compare((ArrayList<Object>)left.get(i), r);
                if (compareResult != 0) return compareResult;
            } else {
                ArrayList<Object> l = new ArrayList<>();
                l.add(left.get(i));
                int compareResult = compare(l, (ArrayList<Object>)right.get(i));
                if (compareResult != 0) return compareResult;
            }
        }
        if (left.size() < right.size()) {
            return -1;
        } else if (right.size() < left.size()) {
            return 1;
        } else {
            return 0;
        }
    }

    public static void main(String[] args) throws FileNotFoundException {
        List<String> lines = new ArrayList<>();

        File inputFile = new File("./input.txt");
        
        Scanner in = new Scanner(inputFile);

        while (in.hasNext()) {
            lines.add(in.nextLine().trim());
        }

        in.close();

        int result = 0;

        int i = 0;

        int pairIndex = 1;

        while (i < lines.size()) {
            String firstLine  = lines.get(i);
            i += 1;
            String secondLine = lines.get(i);
            i += 2;

            ArrayList<Object> firstPacket = constructListFromString(firstLine);
            ArrayList<Object> secondPacket = constructListFromString(secondLine);

            if (compare(firstPacket, secondPacket) == -1) {
                result += pairIndex;
            }

            pairIndex++;
        }

        System.out.println(result);
    }
}